trust-quorum/src/rack_secret_loader.rs

```diff
@@ -0,0 +1,564 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! The component responsible for loading rack secrets for various
+//! configurations.
+
+use std::collections::BTreeMap;
+
+use crate::crypto::ReconstructedRackSecret;
+use crate::{
+    Alarm, Configuration, Epoch, NodeHandlerCtx, PeerMsgKind, PlatformId,
+    RackSecret, Share,
+};
+use slog::{Logger, error, info, o};
+
+#[derive(Debug, Clone, thiserror::Error, PartialEq, Eq)]
+pub enum LoadRackSecretError {
+    #[error("no committed configurations")]
+    NoCommittedConfigurations,
+    #[error("configuration at epoch {0} is not committed")]
+    NotCommitted(Epoch),
+    #[error("alarm: configuration went backwards")]
+    Alarm,
+    #[error("rack secret for epoch {0} is no longer available")]
+    NotAvailable(Epoch),
+}
+
+/// Manage retrieval of key shares to load various rack secrets
+#[derive(Debug, Clone)]
+pub struct RackSecretLoader {
+    log: Logger,
+    loaded: BTreeMap<Epoch, ReconstructedRackSecret>,
+    // We can only collect shares for the latest committed epoch. We then derive
+    // a key from the computed rack secret to decrypt rack secrets for prior
+    // configurations.
+    collector: Option<ShareCollector>,
+}
+
+impl RackSecretLoader {
+    pub fn new(log: &Logger) -> RackSecretLoader {
+        let log = log.new(o!("component" => "tq-rack-secret-loader"));
+        RackSecretLoader { log, loaded: BTreeMap::new(), collector: None }
+    }
+
+    pub fn load(
+        &mut self,
+        ctx: &mut impl NodeHandlerCtx,
+        epoch: Epoch,
+    ) -> Result<Option<ReconstructedRackSecret>, LoadRackSecretError> {
+        // Return the rack secret if we have it
+        if let Some(secret) = self.loaded.get(&epoch) {
+            return Ok(Some(secret.clone()));
+        }
+
+        let Some(latest_committed_epoch) =
+            ctx.persistent_state().latest_committed_epoch()
+        else {
+            // There aren't any committed configs
+            return Err(LoadRackSecretError::NoCommittedConfigurations);
+        };
+
+        // If we have loaded the latest committed epoch, then we have loaded all
+        // possible rack secrets. Secrets for prior epochs are unavailable.
+        if self.loaded.contains_key(&latest_committed_epoch) {
+            if epoch < latest_committed_epoch {
+                return Err(LoadRackSecretError::NotAvailable(epoch));
+            } else if epoch > latest_committed_epoch {
+                return Err(LoadRackSecretError::NotCommitted(epoch));
+            } else {
+                unreachable!(
+                    "already would have returned rack secret for latest \
+                    committed epoch if requested"
+                );
+            }
+        }
+
+        // We don't have the rack secret for the latest committed epoch yet. Are
+        // we still collecting shares?
+        match &mut self.collector {
```

Comment:
I may have to re-order some of this logic with what's above. Certain cases like if we are still collecting for a newer epoch than the latest committed one (the alarm case below) will never be hit if the latest committed epoch is already loaded.

While that alarm case is a bug that would only be hit in the case of corruption to the persistent state (in memory at runtime?) we still want to alert on it if it occurs.